File: examples/canonical-terminal/elixir-optimus-ratatouille/lib/taskflow/core/stats.ex
defmodule Taskflow.Core.Stats do
  @statuses ~w[pending running done failed]

  def compute(jobs) do
    counts = Enum.into(@statuses, %{}, &{&1, 0})

    {counts, durations} =
      Enum.reduce(jobs, {counts, []}, fn job, {acc, durations} ->
        {
          Map.update!(acc, job.status, &(&1 + 1)),
          if(job.duration_s, do: [job.duration_s | durations], else: durations)
        }
      end)

    total = length(jobs)

    %{
      total: total,
      by_status: counts,
      avg_duration_s:
        case durations do
          [] -> 0.0
          _ -> Float.round(Enum.sum(durations) / length(durations), 2)
        end,
      failure_rate: if(total == 0, do: 0.0, else: Float.round(Map.fetch!(counts, "failed") / total, 2))
    }
  end
end

File: examples/canonical-terminal/ruby-clamp-tty/lib/taskflow/core/stats.rb
# frozen_string_literal: true

module Taskflow
  module Core
    module Stats
      module_function

      def compute(jobs)
        counts = %w[pending running done failed].to_h { |status| [status, 0] }
        durations = []

        jobs.each do |job|
          counts[job.status] += 1 if counts.key?(job.status)
          durations << job.duration_s if job.duration_s
        end

        total = jobs.length
        {
          total: total,
          by_status: counts,
          avg_duration_s: durations.empty? ? 0.0 : (durations.sum / durations.length).round(2),
          failure_rate: total.zero? ? 0.0 : (counts["failed"].to_f / total).round(2)
        }
      end
    end
  end
end
